lib/foreman_maintain/utils/bash.rb

```diff
@@ -0,0 +1,138 @@
+require 'yaml'
+
+module ForemanMaintain
+  module Utils
+    module Bash
+      class Completion
+        def initialize(dict)
+          @dict = dict
+        end
+
+        def complete(line)
+          @complete_line = line.end_with?(' ')
+          full_path = line.split(' ')
+          complete_path = @complete_line ? full_path : full_path[0..-2]
+          dict, path = traverse_tree(@dict, complete_path)
+
+          return [] unless path.empty? # lost during traversing
+
+          if @complete_line
+            next_word(dict)
+          else
+            finish_word(dict, full_path.last)
+          end
+        end
+
+        def self.load_description(path)
+          YAML.load(File.open(path))
+        rescue Errno::ENOENT
+          {}
+        end
+
+        private
+
+        def finish_word(dict, incomplete)
+          finish_option_value(dict, incomplete) ||
+            (finish_option_or_subcommand(dict, incomplete) + finish_params(dict, incomplete))
+        end
+
+        def finish_option_or_subcommand(dict, incomplete)
+          dict.keys.select { |k| k.is_a?(String) && k =~ /^#{incomplete}/ }.map { |k| k + ' ' }
+        end
+
+        def next_word(dict)
+          next_option_value(dict) || (next_option_or_subcommand(dict) + next_param(dict))
+        end
+
+        def next_option_or_subcommand(dict)
+          dict.keys.select { |k| k.is_a?(String) }.map { |k| k + ' ' }
+        end
+
+        def complete_value(value, partial, is_param)
```

Comment:
the `value` name is misleading: what about something like `completion_options` instead?

Reply:
how about `value_description`?